Source/Models/APIClient.swift

```diff
@@ -0,0 +1,30 @@
+import Foundation
+import Argo
+import Result
+
+public struct APIClient {
+  private let requestPerformer: RequestPerformer
+
+  public init(requestPerformer: RequestPerformer = NetworkRequestPerformer()) {
+    self.requestPerformer = requestPerformer
+  }
+}
+
+public extension APIClient {
+  func performRequest<T: Request>(request: T, completionHandler: Result<T.ResponseType, NSError> -> Void) {
+    requestPerformer.performRequest(request.build()) { result in
+      let object = result >>- deserialize >>- { request.parse($0) }
+      completionHandler(object)
+    }
+  }
+}
+
+private func deserialize(response: HTTPResponse) -> Result<JSON, NSError> {
+  do {
+    let data = response.data ?? NSData()
```

Comment:
defaulting to `NSData()` will cause the `NSJSONSerialization` process below to fail. That will cause it to throw an error, in which case we will transform it into a `Failure` and return it. I'm not sure if this belongs here, or maybe on `HTTPResponse`. I feel like it should be here, because it seems reasonable that `HTTPResponse` might not have any actual data, and so should represent that with a proper type instead of hiding it.